app/controllers/sessions_controller.rb

```diff
@@ -7,14 +7,46 @@ class SessionsController < Clearance::SessionsController
   def create
     @user = find_user
 
-    if @user&.mfa_enabled?
-      session[:mfa_user] = @user.id
-      render "sessions/otp_prompt"
+    if @user && (@user.mfa_enabled? || @user.webauthn_credentials.any?)
+      setup_webauthn_authentication
+      setup_mfa_authentication
+
+      render "sessions/prompt"
     else
       do_login
     end
   end
 
+  def webauthn_create
+    @user = User.find(session.dig(:webauthn_authentication, "user"))
+    @challenge = session.dig(:webauthn_authentication, "challenge")
+
+    if params[:credentials].blank?
+      login_failure("Credentials required")
+      return
+    end
+
+    @credential = WebAuthn::Credential.from_get(params[:credentials])
+
+    @webauthn_credential = @user.webauthn_credentials.find_by(
+      external_id: @credential.id
+    )
+
+    @credential.verify(
+      @challenge,
+      public_key: @webauthn_credential.public_key,
+      sign_count: @webauthn_credential.sign_count
+    )
+
+    @webauthn_credential.update!(sign_count: @credential.sign_count)
+
+    do_login
+  rescue WebAuthn::Error => e
+    login_failure(e.message)
```

Comment:
@jenshenny Actually, this is a bit finicky since sometimes the error name is included and sometimes it isn't, trying to bypass it causes errors without names to not show up properly (the error you noticed [here](https://github.com/rubygems/rubygems.org/pull/2865#discussion_r1031677309)) . I can do an `if/else` here to exclude the name if it's there but that muddies the code for what (to me) seems like a very minimal benefit. I'd prefer to just leave this with the error name in the response. 

<img width="583" alt="image" src="https://user-images.githubusercontent.com/8496209/204673013-a6a8a771-3e41-48ec-aa75-6697c731a4a1.png">